helper.go

```diff
@@ -107,6 +120,77 @@ func InitHelper(config *Config) (*Helper, error) {
 	}, nil
 }
 
+// symlinkAuxiliaryProviders discovers auxiliary provider binaries, used in
+// multi-provider tests, and symlinks them to the plugin directory.
+//
+// Auxiliary provider binaries should be included in the provider source code
+// directory, under the path terraform.d/plugins/$GOOS_$GOARCH/provider-name.
+//
+// The environment variable TF_ACC_PROVIDER_ROOT_DIR must be set to the path of
+// the provider source code directory root in order to use this feature.
+func symlinkAuxiliaryProviders(pluginDir string) error {
+	providerRootDir := os.Getenv("TF_ACC_PROVIDER_ROOT_DIR")
+	if providerRootDir == "" {
+		// common case; assume intentional and do not log
+		return nil
+	}
+
+	_, err := os.Stat(filepath.Join(providerRootDir, "terraform.d", "plugins"))
+	if err != nil {
+		fmt.Printf("No terraform.d/plugins directory found: continuing. Unset TF_ACC_PROVIDER_ROOT_DIR or supply provider binaries in terraform.d/plugins/$GOOS_$GOARCH to disable this message.")
+		return nil
+	}
+
+	auxiliaryProviderDir := filepath.Join(providerRootDir, "terraform.d", "plugins", runtime.GOOS+"_"+runtime.GOARCH)
+
+	// If we can't os.Stat() terraform.d/plugins/$GOOS_$GOARCH, however,
+	// assume the omission was unintentional, and error.
+	_, err = os.Stat(auxiliaryProviderDir)
+	if err != nil {
+		return fmt.Errorf("error finding auxiliary provider dir %s: %s", auxiliaryProviderDir, err)
+	}
+
+	// now find all the providers in that dir and symlink them to the plugin dir
+	providers, err := ioutil.ReadDir(auxiliaryProviderDir)
+	if err != nil {
+		return fmt.Errorf("error reading auxiliary providers: %s", err)
+	}
+
+	zipDecompressor := new(getter.ZipDecompressor)
+
+	for _, provider := range providers {
+		providerFilename := provider.Name()
+		providerNameExt := filepath.Ext(providerFilename)
+		providerName := strings.TrimSuffix(providerFilename, providerNameExt)
+		providerPath := filepath.Join(auxiliaryProviderDir, providerName)
+		providerSymlinkPath := filepath.Join(pluginDir, providerName)
+
+		// exit early if we have already symlinked this provider
+		_, err := os.Stat(providerSymlinkPath)
+		if err == nil {
+			continue
+		}
+
+		// if filename ends in .zip, assume it is a zip and extract it
+		// otherwise assume it is a provider binary
+		if providerNameExt == ".zip" {
+			_, err = os.Stat(providerPath)
+			if err == nil {
```

Comment:
I would check `os.IsNotExist(err)`, do the decompress, `else if err != nil` return unexpected error, omit `else`.